mantis-network/src/test/java/io/reactivex/mantis/network/push/TimedChunkerTest.java

```diff
@@ -0,0 +1,189 @@
+
+        List<Integer> expected = new ArrayList<>();
+        for (int i = 0; i < 50; i++) {
+            expected.add(i);
+        }
+        ExecutorService service = Executors.newFixedThreadPool(2);
+        Future<Void> chunkerFuture = service.submit(timedChunker);
+        // Send 50 events in 100ms interval. Should take 5s total to send.
+        Future senderFuture = service.submit(() -> {
+            for (int i : expected) {
+                logger.info("sending event {}", i);
+                monitoredQueue.write(i);
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ex) {
+                    logger.info("sender interrupted", ex);
+                }
+            }
+        });
+        Thread.sleep(6000);
+        assertTrue(senderFuture.isDone());
+        chunkerFuture.cancel(true);
+        assertEquals(expected, processor.getProcessed());
+    }
+
+    @Test
+    public void testLongProcessing() throws Exception {
+        // Processing time take longer than drain interval.
+        processor = new TestProcessor<>(400);
+        TimedChunker<Integer> timedChunker = new TimedChunker<>(monitoredQueue, 100, 200, processor, null);
+
+        List<Integer> expected = new ArrayList<>();
+        for (int i = 0; i < 50; i++) {
+            expected.add(i);
+        }
+        ExecutorService service = Executors.newFixedThreadPool(2);
+        Future<Void> chunkerFuture = service.submit(timedChunker);
+        // Send 50 events in 100ms interval. Should take 5s total to send.
+        Future senderFuture = service.submit(() -> {
+            for (int i : expected) {
+                logger.info("sending event {}", i);
+                monitoredQueue.write(i);
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ex) {
+                    logger.info("sender interrupted", ex);
+                }
+            }
+        });
+        Thread.sleep(6000);
+        assertTrue(senderFuture.isDone());
+        chunkerFuture.cancel(true);
+        assertEquals(expected, processor.getProcessed());
+    }
+
+    public static class TestProcessor<T> extends ChunkProcessor<T> {
+        private ScheduledExecutorService scheduledService = Executors.newSingleThreadScheduledExecutor();
+        private List<T> processed = new ArrayList();
+        private long processingTimeMs = 0;
+
+        public TestProcessor(long processingTimeMs) {
+            super(null);
+            this.processingTimeMs = processingTimeMs;
+        }
+
+        @Override
+        public void process(ConnectionManager<T> connectionManager, List<T> chunks) {
+            if (processingTimeMs > 0) {
+                ScheduledFuture<Boolean> f = scheduledService.schedule(() -> true, processingTimeMs,
+                        TimeUnit.MILLISECONDS);
+                    while (!f.isDone()) {
+                    }
+            }
```

Comment:
Redundent?

Reply:
This is needed to emulate long processing time. And we cannot just use Thread.sleep either, we need the thread to keep running.